docs/config.md
# Configuring Will

Most of will's configuration is done interactively, using `run_will.py`, or specifying via the plugins.  There are, however a few built-in settings and config worth covering.  We'll aim to address all of them here.

## Environment variables

All environment variables prefixed with `WILL_` are imported into will's `settings` modules.

In best practices, you should keep all of the following in environment variables:

- `WILL_SLACK_API_TOKEN`
- `WILL_HIPCHAT_USERNAME`
- `WILL_HIPCHAT_PASSWORD`
- `WILL_HIPCHAT_V2_TOKEN`
- `WILL_HIPCHAT_V1_TOKEN`
- `WILL_ROCKETCHAT_USERNAME`
- `WILL_ROCKETCHAT_PASSWORD`
- `WILL_ROCKETCHAT_URL`
- `WILL_REDIS_URL`
- Any other tokens, keys, passwords, or sensitive URLS.

We've made it easy.  No excuses. :)

## config.py

Config.py is where all of your non-sensitive settings should go.   This includes things like:

- `PLUGINS`: The list of plugins to run,
- `PLUGIN_BLACKLIST`: The list of plugins to ignore, even if they're in `PLUGINS`,
- `IO_BACKENDS`: The list services you want Will to connect to,
- `ANALYZE_BACKENDS`: The list of message-analysis backends you want Will to run through.
- `GENERATION_BACKENDS`: The list of reply-generation backends you want Will to go through.
- `EXECUTION_BACKENDS`: The list of decision-making and execution backends you want Will to go through (we recommend just having one.)
- `STORAGE_BACKEND`: Which backend you'd like to use for Will to store his long-term memory. (Built-in: 'redis', 'couchbase', 'file')
- `PUBSUB_BACKEND`: Which backend you'd like to use for Will to use for his working memory. (Built-in: 'redis'.  Soon: 'zeromq', 'builtin')
- `ENCYPTION_BACKEND`: Which backend you'd like to use for Will to encrypt his storage and memory. (Built-in: 'aes'.)
- `PUBLIC_URL`: The publicly accessible URL will can reach himself at (used for [keepalive](plugins/bundled.md#administration)),
- `HTTPSERVER_PORT`: The port will should handle HTTP requests on.  Defaults to 80, set to > 1024 if you don't have sudo,
- `REDIS_MAX_CONNECTIONS`: The maximum number of connections to make to redis, for connection pooling.
- `FUZZY_MINIMUM_MATCH_CONFIDENCE`:  What percentage of confidence Will should have before replying to a fuzzy match.
- `FUZZY_REGEX_ALLOWABLE_ERRORS`:  The maximum number of letters that can be wrong in trying to make a fuzzy match.
- `SLACK_DEFAULT_CHANNEL`: The default Slack channel to send messages to (via webhooks, etc)
- `HIPCHAT_ROOMS`: The list of rooms to join,
- `HIPCHAT_DEFAULT_ROOM`: The room to send messages that come from web requests to,
- `DEFAULT_BACKEND`: The service to send messages that come from web requests to,
- `TEMPLATE_DIRS`: Extra directories to look for templates,
- `ADMINS`: The mention names of all the admins,
- `LOGLEVEL`: What logging level to use,
- `HIPCHAT_SERVER`: if you're using the [HipChat server beta](https://www.hipchat.com/server), the hostname of the server,
- `ALLOW_INSECURE_HIPCHAT_SERVER`: the option to disable SSL checks (seriously, don't),
- `ENABLE_INTERNAL_ENCRYPTION`: the option to turn off internal encryption (not recommended, but you can do it.)
- `PROXY_URL`: Proxy server to use, consider exporting it as `WILL_PROXY_URL` environment variable, if it contains sensitive information
- `ACL`: Define arbitrary groups of users which can be used to restrict access to certain Will commands. See [access control](plugins/builtins/#access-control) for details.
- `IGNORED_USERS`: A list of Slack userids which should be ignored.  Particularly useful to prevent bots from talking to each other.  Slack only.
- and all of your non-sensitive plugin settings.


More expansive documentation on all of those settings is in `config.py`, right where you need it.

## How environment variables and config.py are combined

The environment variables and config.py are combined, and made available to the rest of the app at:

```python
from will import settings

print settings.MY_SETTING_NAME
```

The rules for combining are fairly straightforward:

1. All environment variables that start with `WILL_` are imported, and `WILL_` is stripped off their name. (i.e. `WILL_PORT` becomes `PORT`)
2. All variables from `config.py` are imported.  If there is a conflict, `config.py` wins, and a message is displayed:

    ![Config Conflict](img/config_conflict.gif)

3. Some smart defaulting happens inside settings.py for important variables.  For the moment, I'm going to leave that out of the docs, and refer you to `settings.py` as I *believe* things should Just Work, and most people should never need to care.  If this decision's wrong, please open an issue, and these docs will be improved!

That's it for config.  Now, you can either do a deeper dive into [Will's brain](/backends/overall.md), or just get your will [deployed](deploy.md)!





docs/releases.md
# Releases

#### 2.1.3 - June 13, 2018

Bugfix & feature release that includes:

Bugfixes: 
* Will is now robust against slack disconnects, and automatically handles reconnects, thanks to [https://github.com/Ashex](Ashex)!
* Will also now doesn't respond to every single message when he joins a new channel. Please send your thank-you cards to [https://github.com/Ashex](Ashex). 
* Makes will more robust at handling incorrect channel names in slack, thanks to [reist](https://github.com/reist).
* Properly renames `SLACK_DEFAULT_ROOM` to `SLACK_DEFAULT_CHANNEL` (with backwards-compatability) thanks (again!) to [reist](https://github.com/reist).
* Enable/disable backends from `generate_will_project` now properly puts the comment outside the strong.  Thanks to [phiro69](https://github.com/phiro69) for the report.


New features:
* Adds support for slack attachments, thanks to [https://github.com/Ashex](Ashex)
* Reminders now notify the person making the reminder, thanks to [unicolet](https://github.com/unicolet).
* `generate_will_project` now supports a `--backends` flag, thanks to [wontonst](https://github.com/wontonst).
* Will now explicitly notes if he's automagically using `REDIS_URL` to find the redis backend.

#### 2.1.2 - March 30, 2018

Bugfix release that includes:

* Fixes python 2 compatability (`str` instead of `basestring`) for the HipChat adapter, thanks to [wontonst](https://github.com/wontonst).


#### 2.1.1 - March 22, 2018

Bugfix release that includes:

* Fixes slack reconnect issues, thanks to [@mattcl](https://github.com/mattcl).  Props to [@cmachine](https://github.com/cmachine) for also submitting a fix.
* Saying "G" will no longer give you a picture of a pug, using the default settings. This is both tragic, and necessary. (Actual fix: adjusted default fuziness settings.  If you have the fuzzy backend on, and were seeing the rather hilarious/annoying [#327](https://github.com/skoczen/will/issues/327), set `FUZZY_MINIMUM_MATCH_CONFIDENCE = 91` in your config.py)
* Programmer help is working again, thanks to [@acommasplice](https://github.com/acommasplice).
* Fixes word game to work in python 3 thanks to [@netjunki](https://github.com/netjunki), and [ptomkiel-oktawave](https://github.com/ptomkiel-oktawave)'s report.
* Fixes up chat room rosters in HipChat with rosters > 1000 rooms, thanks to [@ostracon](https://github.com/ostracon)
* Fixes `get_room_from_message`, thanks to [@TaunoTinits](https://github.com/TaunoTinits)'s fix and  [ptomkiel-oktawave](https://github.com/ptomkiel-oktawave)'s report.
* Fixes an error that could occur on incoming webhooks on hipchat.  Thanks to [ptomkiel-oktawave](https://github.com/ptomkiel-oktawave) and others for a report.
* Fixes Will incorrectly talking to the main slack room, when he's directly addressed in 1-1 with something he doesn't know how to do.  Thanks to [netjunki](https://github.com/netjunki) for the report!


#### 2.1.0 - November 28, 2017

Planned release that includes:

* Automatic [docker hub builds](https://hub.docker.com/r/heywill/will/) thanks to [@mike-love](https://github.com/mike-love)
* Upgrades to use base markdownify package, as the proposed changes have been [merged and released](https://github.com/matthewwithanm/python-markdownify/pull/1).
* New fabric commands to manage docker builds and releases.


#### 2.0.2 - November 22, 2017

Bugfix release that fixes:

* Will once again joins all hipchat rooms `HIPCHAT_ROOMS` was not specified.  Thanks to [vissree](https://github.com/vissree) for finding and reporting this bug!

#### 2.0.1 - November 21, 2017

Same release as 2.0.1, removes beta tag.

#### 2.0.1beta4 - November 20, 2017

Bugfix release that fixes:

* `color` parameter wasn't working properly in Slack.
* Fixes up slack escaping, to support `<slack formatted|https://example.com>` links.  Thanks to [@netjunki](https://github.com/netjunki) for the report on this and the above.

Minor features:
* Adds a new `start_thread` parameter to `say()` and `reply()` to allow Will to start slack threads.


#### 2.0.1beta3 - November 13, 2017

Bugfix release that fixes:

* High CPU in some setups, thanks to [mattcl](https://github.com/mattcl) for the report and debuggging!
* Updates to `markdownify` fork 0.4.1.



#### 2.0.1beta2 - November 9, 2017

Bugfix release that fixes:

* Fixes `scheduled_say` breakage.
* Improves reminder plugins to capture and naturally handle "to"s, thanks to [wontonst](https://github.com/wontonst).
* Gets docker builds working, thanks to [mike-love](https://github.com/mike-love).


#### 2.0.1beta1 - November 7, 2017

**TL;DR: Slack, Rocket.chat, and Shell support, and you can write full chatterbots with Will now!**

This is a huge rewrite of will, adding pluggable backends for chat systems, Will's internal brains, pub-sub, and encryption.  

A huge number of really smart people gave their thoughts and suggestions throughout the process, not least [@hobson](https://github.com/hobson), [@woohgit](https://github.com/woohgit), [@netjunki](https://github.com/netjunki), [@sivy](https://github.com/sivy), [@antgel](https://github.com/antgel), [@shadow7412](https://github.com/shadow7412), [@brandonsturgeon](https://github.com/brandonsturgeon), [@pepedocs](https://github.com/pepedocs), [@tophsic](https://github.com/tophsic), and [@mike-love](https://github.com/mike-love).

Read all about what and why here: [What's new in Will 2](https://heywill.io/will2),


will/settings.py
import os
import sys
from will.utils import show_valid, warn, note, error
from clint.textui import puts, indent
from six.moves.urllib import parse
from six.moves import input


def auto_key():
    """This method attempts to auto-generate a unique cryptographic key based on the hardware ID.
    It should *NOT* be used in production, or to replace a proper key, but it can help get will
    running in local and test environments more easily."""
    import uuid
    import time
    import random
    import hashlib

    node = uuid.getnode()

    h = hashlib.md5()
    h.update(str("%s" % node).encode('utf-8'))
    key1 = h.hexdigest()

    time.sleep(random.uniform(0, 0.5))
    node = uuid.getnode()

    h = hashlib.md5()
    h.update(str("%s" % node).encode('utf-8'))
    key2 = h.hexdigest()

    time.sleep(random.uniform(0, 0.5))
    node = uuid.getnode()

    h = hashlib.md5()
    h.update(str("%s" % node).encode('utf-8'))
    key3 = h.hexdigest()

    if key1 == key2 and key2 == key3:
        return key1

    return False


def import_settings(quiet=True):
    """This method takes care of importing settings from the environment, and config.py file.

    Order of operations:
    1. Imports all WILL_ settings from the environment, and strips off the WILL_
    2. Imports settings from config.py
    3. Sets defaults for any missing, required settings.

    This method takes a quiet kwarg, that when False, prints helpful output. Called that way during bootstrapping.
    """

    settings = {}

    # Import from environment, handle environment-specific parsing.
    for k, v in os.environ.items():
        if k[:5] == "WILL_":
            k = k[5:]
            settings[k] = v
    if "HIPCHAT_ROOMS" in settings and type(settings["HIPCHAT_ROOMS"]) is type("tes"):
        settings["HIPCHAT_ROOMS"] = settings["HIPCHAT_ROOMS"].split(";")

    if "ROOMS" in settings:
        settings["ROOMS"] = settings["ROOMS"].split(";")

    if "PLUGINS" in settings:
        settings["PLUGINS"] = settings["PLUGINS"].split(";")

    if 'PLUGIN_BLACKLIST' in settings:
        settings["PLUGIN_BLACKLIST"] = (settings["PLUGIN_BLACKLIST"].split(";")
                                        if settings["PLUGIN_BLACKLIST"] else [])

    # If HIPCHAT_SERVER is set, we need to change the USERNAME slightly
    # for XMPP to work.
    if "HIPCHAT_SERVER" in settings:
        settings["USERNAME"] = "{user}@{host}".\
            format(user=settings["USERNAME"].split("@")[0],
                   host=settings["HIPCHAT_SERVER"])
    else:
        settings["HIPCHAT_SERVER"] = "api.hipchat.com"

    # Import from config
    if not quiet:
        puts("Importing config.py... ")
    with indent(2):
        try:
            had_warning = False
            try:
                import config
            except ImportError:
                # Missing config.py.  Check for config.py.dist
                if os.path.isfile("config.py.dist"):
                    confirm = input(
                        "Hi, looks like you're just starting up!\nI didn't find a config.py, but I do see config.py.dist here. Want me to use that? (y/n) "
                    ).lower()
                    if confirm in ["y", "yes"]:
                        print("Great! One moment.\n\n")
                        os.rename("config.py.dist", "config.py")
                        import config
                    else:
                        print("Ok.  I can't start without one though. Quitting now!")
                        sys.exit(1)
                else:
                    error("I'm missing my config.py file. Usually one comes with the installation - maybe it got lost?")
                    sys.exit(1)

            for k, v in config.__dict__.items():
                # Ignore private variables
                if "__" not in k:
                    if k in os.environ and v != os.environ[k] and not quiet:
                        warn("%s is set in the environment as '%s', but overridden in"
                             " config.py as '%s'." % (k, os.environ[k], v))
                        had_warning = True
                    settings[k] = v

            if not had_warning and not quiet:
                show_valid("Valid.")
        except:
            # TODO: Check to see if there's a config.py.dist
            if not quiet:
                warn("no config.py found.  This might be ok, but more likely, "
                     "you haven't copied config.py.dist over to config.py")

    if not quiet:
        puts("Verifying settings... ")

    with indent(2):
        # Deprecation and backwards-compatibility for Will 1.x-> 2.x
        DEPRECATED_BUT_MAPPED_SETTINGS = {
            "USERNAME": "HIPCHAT_USERNAME",
            "PASSWORD": "HIPCHAT_PASSWORD",
            "V1_TOKEN": "HIPCHAT_V1_TOKEN",
            "V2_TOKEN": "HIPCHAT_V2_TOKEN",
            "TOKEN": "HIPCHAT_V1_TOKEN",
            "ROOMS": "HIPCHAT_ROOMS",
            "NAME": "HIPCHAT_NAME",
            "HANDLE": "HIPCHAT_HANDLE",
            "DEFAULT_ROOM": "HIPCHAT_DEFAULT_ROOM",
            "SLACK_DEFAULT_ROOM": "SLACK_DEFAULT_CHANNEL",
        }
        deprecation_warn_shown = False
        for k, v in DEPRECATED_BUT_MAPPED_SETTINGS.items():
            if not v in settings and k in settings:
                if not deprecation_warn_shown and not quiet:
                    error("Deprecated settings. The following settings will stop working in Will 2.2:")
                    deprecation_warn_shown = True
                if not quiet:
                    warn("Please update %s to %s.  " % (k, v))
                settings[v] = settings[k]
                del settings[k]

        # Migrate from 1.x
        if "CHAT_BACKENDS" in settings and "IO_BACKENDS" not in settings:
            IO_BACKENDS = []
            for c in settings["CHAT_BACKENDS"]:
                IO_BACKENDS.append("will.backends.io_adapters.%s" % c)
            settings["IO_BACKENDS"] = IO_BACKENDS
            if not quiet:
                warn(
                    "Deprecated settings.  Please update your config.py from:"
                    "\n   CHAT_BACKENDS = %s\n   to\n   IO_BACKENDS = %s" %
                    (settings["CHAT_BACKENDS"], IO_BACKENDS)
                )
        if "CHAT_BACKENDS" not in settings and "IO_BACKENDS" not in settings:
            if not quiet:
                warn("""Deprecated settings.  No backend found, so we're defaulting to hipchat and shell only.
Please add this to your config.py:
IO_BACKENDS = "
    "will.backends.io_adapters.hipchat",
    "will.backends.io_adapters.shell",
#   "will.backends.io_adapters.slack",
#   "will.backends.io_adapters.rocketchat",
]
""")
            settings["IO_BACKENDS"] = [
                "will.backends.io_adapters.hipchat",
                "will.backends.io_adapters.shell",
            ]

        if "ANALYZE_BACKENDS" not in settings:
            if not quiet:
                note("No ANALYZE_BACKENDS specified.  Defaulting to history only.")
            settings["ANALYZE_BACKENDS"] = [
                "will.backends.analysis.nothing",
                "will.backends.analysis.history",
            ]

        if "GENERATION_BACKENDS" not in settings:
            if not quiet:
                note("No GENERATION_BACKENDS specified.  Defaulting to fuzzy_all_matches and strict_regex.")
            settings["GENERATION_BACKENDS"] = [
                "will.backends.generation.fuzzy_all_matches",
                "will.backends.generation.strict_regex",
            ]

        if "EXECUTION_BACKENDS" not in settings:
            if not quiet:
                note("No EXECUTION_BACKENDS specified.  Defaulting to best_score.")
            settings["EXECUTION_BACKENDS"] = [
                "will.backends.execution.best_score",
            ]

        # Set for hipchat
        for b in settings["IO_BACKENDS"]:
            if "hipchat" in b:
                if "ALLOW_INSECURE_HIPCHAT_SERVER" in settings \
                        and (
                        settings["ALLOW_INSECURE_HIPCHAT_SERVER"] is True
                        or settings["ALLOW_INSECURE_HIPCHAT_SERVER"].lower() == "true"
                        ):
                    warn("You are choosing to run will with SSL disabled. "
                         "This is INSECURE and should NEVER be deployed outside a development environment.")
                    settings["ALLOW_INSECURE_HIPCHAT_SERVER"] = True
                    settings["REQUESTS_OPTIONS"] = {
                        "verify": False,
                    }
                else:
                    settings["ALLOW_INSECURE_HIPCHAT_SERVER"] = False

                if "HIPCHAT_ROOMS" not in settings:
                    if not quiet:
                        warn("no HIPCHAT_ROOMS list found in the environment or config.  "
                             "This is ok - Will will just join all available HIPCHAT_rooms.")
                        settings["HIPCHAT_ROOMS"] = None

                if (
                    "HIPCHAT_DEFAULT_ROOM" not in settings and "HIPCHAT_ROOMS" in settings
                    and settings["HIPCHAT_ROOMS"] and len(settings["HIPCHAT_ROOMS"]) > 0
                ):
                    if not quiet:
                        warn("no HIPCHAT_DEFAULT_ROOM found in the environment or config.  "
                             "Defaulting to '%s', the first one." % settings["HIPCHAT_ROOMS"][0])
                    settings["HIPCHAT_DEFAULT_ROOM"] = settings["HIPCHAT_ROOMS"][0]

            if "HIPCHAT_HANDLE" in settings and "HIPCHAT_HANDLE_NOTED" not in settings:
                if not quiet:
                    note(
                        """HIPCHAT_HANDLE is no longer required (or used), as Will knows how to get\n
                                his current handle from the HipChat servers."""
                    )
                    settings["HIPCHAT_HANDLE_NOTED"] = True

            if "HIPCHAT_NAME" in settings and "HIPCHAT_NAME_NOTED" not in settings:
                if not quiet:
                    note(
                        """HIPCHAT_NAME is no longer required (or used), as Will knows how to get\n
                                his current name from the HipChat servers."""
                    )
                    settings["HIPCHAT_NAME_NOTED"] = True

        # Rocket.chat
        for b in settings["IO_BACKENDS"]:
            if "rocketchat" in b:
                if "ROCKETCHAT_USERNAME" in settings and "ROCKETCHAT_EMAIL" not in settings:
                    settings["ROCKETCHAT_EMAIL"] = settings["ROCKETCHAT_USERNAME"]
                if "ROCKETCHAT_URL" in settings:
                    if settings["ROCKETCHAT_URL"].endswith("/"):
                        settings["ROCKETCHAT_URL"] = settings["ROCKETCHAT_URL"][:-1]

        if (
            "DEFAULT_BACKEND" not in settings and "IO_BACKENDS" in settings
            and settings["IO_BACKENDS"] and len(settings["IO_BACKENDS"]) > 0
        ):
            if not quiet:
                note("no DEFAULT_BACKEND found in the environment or config.\n  "
                     "      Defaulting to '%s', the first one." % settings["IO_BACKENDS"][0])
            settings["DEFAULT_BACKEND"] = settings["IO_BACKENDS"][0]

        for b in settings["IO_BACKENDS"]:
            if "slack" in b and "SLACK_DEFAULT_CHANNEL" not in settings and not quiet:
                warn(
                    "No SLACK_DEFAULT_CHANNEL set - any messages sent without an explicit channel will go "
                    "to a non-deterministic channel that will has access to "
                    "- this is almost certainly not what you want."
                )

        if "HTTPSERVER_PORT" not in settings:
            # For heroku
            if "PORT" in os.environ:
                settings["HTTPSERVER_PORT"] = os.environ["PORT"]
            else:
                if not quiet:
                    warn("no HTTPSERVER_PORT found in the environment or config.  Defaulting to ':80'.")
                settings["HTTPSERVER_PORT"] = "80"

        if "STORAGE_BACKEND" not in settings:
            if not quiet:
                warn("No STORAGE_BACKEND specified.  Defaulting to redis.")
            settings["STORAGE_BACKEND"] = "redis"

        if "PUBSUB_BACKEND" not in settings:
            if not quiet:
                warn("No PUBSUB_BACKEND specified.  Defaulting to redis.")
            settings["PUBSUB_BACKEND"] = "redis"

        if settings["STORAGE_BACKEND"] == "redis" or settings["PUBSUB_BACKEND"] == "redis":
            if "REDIS_URL" not in settings:
                # For heroku
                if "REDIS_URL" in os.environ:
                    settings["REDIS_URL"] = os.environ["REDIS_URL"]
                    if not quiet:
                        note("WILL_REDIS_URL not set, but it appears you're using Heroku Redis or another standard REDIS_URL. If so, all good.")
                if "REDISCLOUD_URL" in os.environ:
                    settings["REDIS_URL"] = os.environ["REDISCLOUD_URL"]
                    if not quiet:
                        note("WILL_REDIS_URL not set, but it appears you're using RedisCloud. If so, all good.")
                elif "REDISTOGO_URL" in os.environ:
                    settings["REDIS_URL"] = os.environ["REDISTOGO_URL"]
                    if not quiet:
                        note("WILL_REDIS_URL not set, but it appears you're using RedisToGo. If so, all good.")
                elif "OPENREDIS_URL" in os.environ:
                    settings["REDIS_URL"] = os.environ["OPENREDIS_URL"]
                    if not quiet:
                        note("WILL_REDIS_URL not set, but it appears you're using OpenRedis. If so, all good.")
                else:
                    settings["REDIS_URL"] = "redis://localhost:6379/7"
                    if not quiet:
                        note("WILL_REDIS_URL not set.  Defaulting to redis://localhost:6379/7.")

            if not settings["REDIS_URL"].startswith("redis://"):
                settings["REDIS_URL"] = "redis://%s" % settings["REDIS_URL"]

            if "REDIS_MAX_CONNECTIONS" not in settings or not settings["REDIS_MAX_CONNECTIONS"]:
                settings["REDIS_MAX_CONNECTIONS"] = 4
                if not quiet:
                    note("REDIS_MAX_CONNECTIONS not set. Defaulting to 4.")

        if settings["STORAGE_BACKEND"] == "file":
            if "FILE_DIR" not in settings:
                settings["FILE_DIR"] = "~/.will/"
                if not quiet:
                    note("FILE_DIR not set.  Defaulting to ~/.will/")

        if settings["STORAGE_BACKEND"] == "couchbase":
            if "COUCHBASE_URL" not in settings:
                settings["COUCHBASE_URL"] = "couchbase:///will"
                if not quiet:
                    note("COUCHBASE_URL not set.  Defaulting to couchbase:///will")

        if "PUBLIC_URL" not in settings:
            default_public = "http://localhost:%s" % settings["HTTPSERVER_PORT"]
            settings["PUBLIC_URL"] = default_public
            if not quiet:
                note("no PUBLIC_URL found in the environment or config.\n        Defaulting to '%s'." % default_public)

        if not "REQUESTS_OPTIONS" in settings:
            settings["REQUESTS_OPTIONS"] = {}

        if "TEMPLATE_DIRS" not in settings:
            if "WILL_TEMPLATE_DIRS_PICKLED" in os.environ:
                # All good
                pass
            else:
                settings["TEMPLATE_DIRS"] = []

        if "WILL_HANDLE" not in settings:
            if "HANDLE" in settings:
                settings["WILL_HANDLE"] = settings["HANDLE"]
            elif "SLACK_HANDLE" in settings:
                settings["WILL_HANDLE"] = settings["SLACK_HANDLE"]
            elif "HIPCHAT_HANDLE" in settings:
                settings["WILL_HANDLE"] = settings["HIPCHAT_HANDLE"]
            elif "ROCKETCHAT_HANDLE" in settings:
                settings["WILL_HANDLE"] = settings["ROCKETCHAT_HANDLE"]
            else:
                settings["WILL_HANDLE"] = "will"

        if "ADMINS" not in settings:
            settings["ADMINS"] = "*"
        else:
            if "WILL_ADMINS" in os.environ:
                settings["ADMINS"] = [a.strip().lower() for a in settings.get('ADMINS', '').split(';') if a.strip()]

        if "ADMINS" in settings and settings["ADMINS"] != "*":
            warn("ADMINS is now deprecated, and will be removed at the end of 2017.  Please use ACL instead. See below for details")
            note("Change your config.py to:\n  ACL = {\n     'admins': %s\n  }" % settings["ADMINS"])

        if "DISABLE_ACL" not in settings:
            settings["DISABLE_ACL"] = False

        if "PROXY_URL" in settings:
            parsed_proxy_url = parse.urlparse(settings["PROXY_URL"])
            settings["USE_PROXY"] = True
            settings["PROXY_HOSTNAME"] = parsed_proxy_url.hostname
            settings["PROXY_USERNAME"] = parsed_proxy_url.username
            settings["PROXY_PASSWORD"] = parsed_proxy_url.password
            settings["PROXY_PORT"] = parsed_proxy_url.port
        else:
            settings["USE_PROXY"] = False

        if "EVENT_LOOP_INTERVAL" not in settings:
            settings["EVENT_LOOP_INTERVAL"] = 0.025

        if "LOGLEVEL" not in settings:
            settings["LOGLEVEL"] = "ERROR"

        if "ENABLE_INTERNAL_ENCRYPTION" not in settings:
            settings["ENABLE_INTERNAL_ENCRYPTION"] = True

        if "SECRET_KEY" not in settings:
            if not quiet:
                if "ENABLE_INTERNAL_ENCRYPTION" in settings and settings["ENABLE_INTERNAL_ENCRYPTION"]:
                    key = auto_key()
                    if key:
                        warn(
                            """No SECRET_KEY specified and ENABLE_INTERNAL_ENCRYPTION is on.\n
                               Temporarily auto-generating a key specific to this computer:\n    {}\n
                               Please set WILL_SECRET_KEY in the environment as soon as possible to ensure \n
                               Will is able to access information from previous runs.""".format(key)
                        )
                    else:
                        error(
                            """ENABLE_INTERNAL_ENCRYPTION is turned on, but a SECRET_KEY has not been given.\n
                               We tried to automatically generate temporary SECRET_KEY, but this appears to be a \n"
                               shared or virtualized environment.\n Please set a unique secret key in the
                               environment as WILL_SECRET_KEY to run will."""
                        )
                        print("  Unable to start will without a SECRET_KEY while encryption is turned on. Shutting down.")
                        sys.exit(1)

                    settings["SECRET_KEY"] = key
                    os.environ["WILL_SECRET_KEY"] = settings["SECRET_KEY"]
                    os.environ["WILL_EPHEMERAL_SECRET_KEY"] = "True"

        if "FUZZY_MINIMUM_MATCH_CONFIDENCE" not in settings:
            settings["FUZZY_MINIMUM_MATCH_CONFIDENCE"] = 91
        if "FUZZY_REGEX_ALLOWABLE_ERRORS" not in settings:
            settings["FUZZY_REGEX_ALLOWABLE_ERRORS"] = 3

        # Set them in the module namespace
        for k in sorted(settings, key=lambda x: x[0]):
            if not quiet:
                show_valid(k)
            globals()[k] = settings[k]


import_settings()
